### PR TITLE
feat[contracts]: Add L2ChugSplashDeployer contract to manage L2 deployments

### LINKED ISSUE #703: ChugSplash needs a contract that authenticates and executes upgrades as internal transactions
**Is your feature request related to a problem? Please describe.**
Deployments at our level of complexity require multiple transactions spaced out over multiple blocks. This introduces annoyances around ensuring that all deployments are carried out to completion before operation of the chain continues.

Additionally, authenticating a deployment should be extremely high security & therefore it is extremely expensive to perform a deployment where you need to authenticate every transaction in a deployment that requires, say, 20 transactions. It's just unreasonable to sign 20 multisig txs!

**Describe the solution you'd like**
In order to solve this we propose a deployment system which:
1. Can authenticate a very large deployment in 1 transaction and allow **anyone** to actually execute the deployment.
2. Locks the entire contract system while a deployment is in progress, ensuring that no transactions are executed against a partial deployment.
3. Can perform deployments across L1 and L2 (the deployment pauses both L1 and L2 contracts so that we can upgrade any component of our system)
4. Will never get stuck in the middle of a deployment.
5. Tracks the version of the deployment.

To create this we need to create a DeploymentManager contract which handles initiating a new upgrade, shutting down the system, performing upgrades, updating the upgrade if needed, resuming the system, and emitting useful information which describes what the upgrade did.

There is a WIP implementation here: https://github.com/ethereum-optimism/optimism/blob/a33bf584e598bc1d9257d2134a79dc9a8f1f4add/packages/contracts/contracts/chugsplash/L2/ChugSplashDeployer.sol#L13

This issue is specifically scoped to implementing the DeploymentManager contract. Note that we will need to extend the functionality when we want to implement the specific pausing & resuming logic for L1 and L2 as they differ per environment.

**Blockers**
https://github.com/ethereum-optimism/optimism/issues/697
https://github.com/ethereum-optimism/optimism/issues/696

### PR DESCRIPTION
**Pull request by smartcontracts**
**Description**
Adds the `L2ChugSplashDeployer` contract to be used to manage L2 contract deployments and upgrades. Needs tests but it's relatively straightforward.

**Metadata**
- Fixes #703

### CODE REVIEW

**Comment by smartcontracts** [comment]
Note that this PR stubs out `ovmSETCODE` and `ovmSETSTORAGE` since that's going to be completed by #793

File: packages/contracts/contracts/chugsplash/L2ChugSplashDeployer.sol [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    )
+        public
+        onlyOwner
+    {
+        require(
+            hasActiveBundle() == false,
+            "ChugSplashDeployer: previous bundle has not yet been fully executed"
+        );
+
+        currentBundleNonce += 1;
```
Comment (smartcontracts) [inline packages/contracts/contracts/chugsplash/L2ChugSplashDeployer.sol:128]: Should this function emit an event?

File: packages/contracts/contracts/chugsplash/L2ChugSplashDeployer.sol [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+     * Constructor *
+     ***************/
+    
+    /**
+     * @param _owner Address that will initially own the L2ChugSplashDeployer.
+     */
+    constructor(
+        address _owner
+    ) {
+        owner = _owner;
```
Comment (smartcontracts) [inline packages/contracts/contracts/chugsplash/L2ChugSplashDeployer.sol:63]: TODO: Add this contract to the state dump process + add a new input variable for "chugsplash owner"

Comment (ben-chain) [reply packages/contracts/contracts/chugsplash/L2ChugSplashDeployer.sol:63]: Should we we add this TODO inline?  Or is it a blocker for merging this PR?

**Review by ben-chain — CHANGES_REQUESTED** [review]
This is shaping up, requested some changes but looking great! 😊 

One overall note on the current functionality is that while we have the ability to cancel upgrades, we do not have the ability to atomically change the bundle to a new upgrade along with cancelling the one in progress.  It seems like for many cases where we suddenly decide to cancel an upgrade, it could be desired to downgrade/undo/change the state in lockstep, i.e., without re-enabling other transactions in between.  Thoughts on this?  Maybe it's worth adding an `overrideTransactionBundle` or having the `cancel` function accept a param for a new bundle.

**Review by maurelian — APPROVED** [review]
Left some comments and nits, but overall it LGTM.

**Comment by maurelian** [comment]
Oh yeah, gentle reminder to [add a changeset](https://github.com/ethereum-optimism/optimism/pull/755#issuecomment-832312675).

File: packages/contracts/contracts/chugsplash/L2ChugSplashDeployer.sol (hunk 2 above)
Comment (smartcontracts) [reply packages/contracts/contracts/chugsplash/L2ChugSplashDeployer.sol:63]: Blocker for merging PR.

**Review by maurelian — APPROVED** [review]
Changes look really good to me. Thanks for adding the docstrings!
FWIW, In this review I considered whether or not it would be possible to get a particular action hash intended to do one thing, to do something else. AFAICT the answer is no, because the action type is included in the hash, which then defines the encoding.

**Review by ben-chain — APPROVED** [review]

**Pull request merged**